Shared/Samples/Create load report/CreateLoadReportView.Model.swift

```diff
@@ -0,0 +1,375 @@
+// Copyright 2023 Esri
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import ArcGIS
+import SwiftUI
+
+extension CreateLoadReportView {
+    /// The model used to store the geo model and other expensive objects
+    /// used in this view.
+    @MainActor
+    class Model: ObservableObject {
+        // MARK: Properties
+        /// The utility network for this sample.
+        private let utilityNetwork = UtilityNetwork(url: .utilityNetwork)
+        
+        /// The initial conditional expression.
+        private var initialExpression: UtilityTraceConditionalExpression?
+        
+        /// The trace parameters for creating load reports.
+        private var traceParameters: UtilityTraceParameters?
+        
+        /// The network attributes for the comparison.
+        private var phasesNetworkAttribute: UtilityNetworkAttribute?
+        
+        /// A list of possible phases populated from the network's attributes.
+        /// By default, they are not included in the load report.
+        @Published private(set) var excludedPhases = [CodedValue]()
+        
+        /// A list of phases that are included in the load report.
+        @Published private(set) var includedPhases = [CodedValue]() {
+            didSet { updateAllowsCreateLoadReport() }
+        }
+        
+        /// A list of possible phases populated from the network's attributes.
+        private var allPhases = [CodedValue]()
+        
+        /// The phase summaries in the load report.
+        private var summaries = PhaseSummaries()
+        
+        /// A Boolean value indicating if the load report can be generated.
+        @Published private(set) var allowsCreateLoadReport = false
+        
+        /// The status text to display to the user.
+        @Published private(set) var statusText: String?
+        
+        /// An error that occurred during setup.
+        private(set) var setupError: Error? {
+            didSet {
+                updateAllowsCreateLoadReport()
+                isShowingSetupError = setupError != nil
+            }
+        }
+        
+        /// A Boolean value indicating whether a setup error is showing.
+        @Published var isShowingSetupError = false
```

Comment:
Next time you can put the error and associated booleans in the view. In theory, the data model shouldn't know anything about the view. Here with the boolean in the model, the model knows the states of the view.

Reply:
> In theory, the data model shouldn't know anything about the view.

I think there are various ways to do these things. One of those ways is MVVM.  In the case of MVVM, the view model doesn't know how the view presents an error, but it knows if the error should be presented.  This makes the view model testable without a UI.

Opinions vary on this. But I'm of the opinion personally that MVVM works well with SwiftUI and that testable view models are nice.

Also in this case, the model had to know about the setup error so that it could enforce the "allows create load report" boolean. That would mean that the view and model would have duplicated state if the setup error was also on the view.

Reply:
agree